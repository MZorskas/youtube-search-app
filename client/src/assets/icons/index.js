import { ReactComponent as CloseIcon } from './symbol_x.svg';
import { ReactComponent as SearchIcon } from './search.svg';

export { CloseIcon, SearchIcon };
