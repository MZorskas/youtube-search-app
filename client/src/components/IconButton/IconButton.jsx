import React from 'react';
import classNames from 'classnames';
import PropTypes from 'prop-types';
import './IconButton.scss';

const IconButton = React.forwardRef(
  ({ icon, onClick, className, type, form, ariaLabel }, ref) => {
    return (
      <button
        className={classNames('icon-button', className)}
        onClick={onClick}
        ref={ref}
        type={type}
        form={form}
        aria-label={ariaLabel}
      >
        {icon}
      </button>
    );
  }
);

IconButton.propTypes = {
  icon: PropTypes.node.isRequired,
  className: PropTypes.string,
  onClick: PropTypes.func,
  type: PropTypes.string,
  form: PropTypes.string,
  ariaLabel: PropTypes.string,
};

IconButton.defaultProps = {
  type: 'button',
  ariaLabel: 'search',
};

export default IconButton;
