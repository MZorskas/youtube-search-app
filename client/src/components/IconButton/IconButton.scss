@import 'src/styles/base/base';

.icon-button {
  align-items: center;
  background-image: linear-gradient(
    to right,
    $color-button-primary 0%,
    darken($color-button-primary, 20%) 51%,
    $color-button-primary 100%
  );
  background-size: 200%;
  border: none;
  box-sizing: border-box;
  color: $color-font-primary;
  cursor: pointer;
  display: flex;
  justify-content: center;
  transition: 0.5s;

  &:hover {
    background-position: right center;
  }

  &:focus {
    background-position: right center;
  }
}
