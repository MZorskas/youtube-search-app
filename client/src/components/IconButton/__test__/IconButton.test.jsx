import React from 'react';
import ReactDom from 'react-dom';
import IconButton from '../IconButton';
import { SearchIcon } from 'assets/icons';
import { render, fireEvent, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';

describe('IconButton component', () => {
  it('renders with propTypes warning', () => {
    const spy = jest.spyOn(global.console, 'error');
    const div = document.createElement('div');
    ReactDom.render(<IconButton />, div);
    expect(spy).toHaveBeenCalled();
  });

  it('renders without crashing', () => {
    const spy = jest.spyOn(global.console, 'error');
    const div = document.createElement('div');
    ReactDom.render(<IconButton icon={<SearchIcon />} />, div);
    expect(spy).not.toHaveBeenCalled();
  });

  it('renders with correct attributes', () => {
    const spy = jest.spyOn(global.console, 'error');
    render(
      <IconButton
        type="button"
        form="form-id"
        ariaLabel="search"
        icon={<SearchIcon />}
      />
    );

    expect(screen.getByTestId('icon-button').getAttribute('type')).toBe(
      'button'
    );
    expect(screen.getByTestId('icon-button').getAttribute('form')).toBe(
      'form-id'
    );
    expect(screen.getByTestId('icon-button').getAttribute('aria-label')).toBe(
      'search'
    );
    expect(spy).not.toHaveBeenCalled();
  });

  it('renders with additional className', () => {
    render(<IconButton className="icon-button-test" icon={<SearchIcon />} />);

    expect(screen.getByTestId('icon-button').className).toBe(
      'icon-button icon-button-test'
    );
  });

  it('fires onClick event', () => {
    const mockOnClick = jest.fn();

    render(<IconButton icon={<SearchIcon />} onClick={mockOnClick} />);

    fireEvent.click(screen.getByTestId('icon-button'));
    expect(mockOnClick).toHaveBeenCalledTimes(1);
  });

  it('matches snapshot', () => {
    const tree = renderer.create(<IconButton icon={<SearchIcon />} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('matches snapshot with additional className', () => {
    const tree = renderer
      .create(<IconButton className="icon-button-test" icon={<SearchIcon />} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('matches snapshot with correct attributes', () => {
    const tree = renderer
      .create(
        <IconButton
          type="button"
          form="form-id"
          ariaLabel="search"
          icon={<SearchIcon />}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
