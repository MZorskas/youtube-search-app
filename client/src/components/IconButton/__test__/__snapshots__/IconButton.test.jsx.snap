// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IconButton component matches snapshot 1`] = `
<button
  aria-label="search"
  className="icon-button"
  data-testid="icon-button"
  type="button"
>
  <svg>
    search.svg
  </svg>
</button>
`;

exports[`IconButton component matches snapshot with additional className 1`] = `
<button
  aria-label="search"
  className="icon-button icon-button-test"
  data-testid="icon-button"
  type="button"
>
  <svg>
    search.svg
  </svg>
</button>
`;

exports[`IconButton component matches snapshot with correct attributes 1`] = `
<button
  aria-label="search"
  className="icon-button"
  data-testid="icon-button"
  form="form-id"
  type="button"
>
  <svg>
    search.svg
  </svg>
</button>
`;
